source/common/http/conn_manager_impl.cc

```diff
@@ -1218,10 +1218,20 @@ void ConnectionManagerImpl::ActiveStream::refreshDurationTimeout() {
   if (disable_timer) {
     if (max_stream_duration_timer_) {
       max_stream_duration_timer_->disableTimer();
+      if (route->usingNewTimeouts() && Grpc::Common::isGrpcRequestHeaders(*request_headers_)) {
+        request_headers_->removeGrpcTimeout();
+      }
     }
     return;
   }
 
+  // Set the header timeout before doing used-time adjustements.
+  // This may result in the upstream not getting the latest results, but also
+  // avoids every request getting a custom timeout based on envoy think time.
+  if (route->usingNewTimeouts() && Grpc::Common::isGrpcRequestHeaders(*request_headers_)) {
```

Comment:
Also noteworthy, the new code does *not* set the grpc-timeout like the old code does.

The old code would set the grpc-timeout to whatever was set in Envoy, but I'm not confident that's the correct behavior.
Imagine here selecting one route which took the header based timeout of 4s and lowered to 2s, then a route change to a route with no timeout.  I think functionally we'd want to set the grpc-timeout header iff Envoy lowered the timeout, which would require latching the original downstream timeout.  Which I can go do if you think it's worth it, but given Envoy doesn't advertise any other timeouts I'm wondering if we can just leave as-is?